File: Data Structures & Algorithms/combination-target-sum/submission-1.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        nums.sort()
        def solve(i,t,temp):
            if t == 0:
                res.append(temp.copy())
                return
            if i == len(nums) or nums[i] > t:
                return 
            temp.append(nums[i])
            solve(i, t-nums[i], temp)
            temp.pop()
            solve(i+1, t, temp)
        solve(0, target, [])
        return res

File: Data Structures & Algorithms/target-sum/submission-1.cpp
class Solution {
public:
    int solve(int index, int target, vector<int> &nums, map<int,map<int,int>> &dp) {
        if (index == nums.size()) {
            return target == 0;
        }
        if (dp.contains(index) && dp[index].contains(target)) return dp[index][target];
        dp[index][target] = solve(index+1, target - nums[index], nums, dp) + solve(index + 1, target+nums[index], nums, dp);
        return dp[index][target];

    }
    int findTargetSumWays(vector<int>& nums, int target) {
        int n = nums.size();
        map<int, map<int,int>> dp;
        return solve(0, target, nums, dp);
    }
};
